### PR TITLE
[Creature/Scripting]: Implement StringId for Creatures, assigning a…

### PR DESCRIPTION
**Pull request by Seyden**
… string in the database and being able to identify a creature via the StringID

**Changes proposed:**

- New StringId Field in the Creature Table
- New StringId Field in the Creature_Template Table
- Expose StringId and StringIdIndex to Scripts (retrieving a uint32 via a string and being able to use this to identify a creature)
- It could be said that this StringId functions like a Guid just for scripting
- Add StringIdIndex Support for FindNearestCreatureWithOptions
- Add GetCreatureListWithOptions
- Add StringIdIndex Support for GetCreatureListWithOptions

**Example on how to use in script:**

**Example 1: How to find a single Creature**
```c
FindCreatureOptions options = FindCreatureOptions().SetStringId("NpcGroup1TrashMob");

if (Creature* creature = me->FindNearestCreatureWithOptions(100.f, options))
    creature->Say("Hello i got the StringId NpcGroup1TrashMob", LANG_UNIVERSAL);
```

**Example 2: How to find a list of Creatures**
```c
FindCreatureOptions options = FindCreatureOptions().SetStringId("NpcGroup1TrashMobs");

std::list<Creature*> creatureList;
GetCreatureListWithOptionsInGrid(creatureList, 100.f, options);
for (Creature* creature : creatureList)
    creature->Say("Hello i got the StringId NpcGroup1TrashMobs", LANG_UNIVERSAL);
```

**Issues addressed:**

Closes #  (insert issue tracker number)

**Tests performed:**

(Does it build, tested in-game, etc.)

- [x] Tested ingame
- [x] Builds

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] No Gameobject support
- [ ] No AreaTrigger support

### CODE REVIEW

File: sql/updates/world/master/XXXX_XX_XX_XX_world.sql [hunk 1]
```diff
@@ -1,5 +1,5 @@
 --
-UPDATE trinity_string SET content_default = 'AIName: %s\r\nScriptName: %s\r\nScriptTag: %s\r\nScriptTagId: %u' WHERE entry = 5031;
+UPDATE trinity_string SET content_default = 'AIName: %s\r\nScriptName: %s\r\StringID: %s\r\StringID Index: %u' WHERE entry = 5031;
 
 ALTER TABLE creature 
-  ADD COLUMN ScriptTag VARCHAR(64) DEFAULT NULL
\ No newline at end of file
+  ADD COLUMN StringId VARCHAR(64) DEFAULT NULL;
```
Comment (Aokromes) [inline sql/updates/world/master/XXXX_XX_XX_XX_world.sql:5]: missing newline

**Comment by malcrom** [comment]
StringId field in creature table is a varchar but GetCreatureByStringId(uint32) in core uses an int value. What am I missing?

File: src/server/game/Globals/ObjectMgr.h [hunk 2]
```diff
@@ -861,6 +861,21 @@ struct SceneTemplate
...
+
+    uint32 Index = 0;
+
+    explicit operator bool() const { return Index > 0; }
+
+    bool operator==(uint32 a) const;
+    bool operator!=(uint32 a) const { return !(operator==(a)); }
+
+    operator uint32() const { return Index; }
+};
```
Comment (Warpten) [inline src/server/game/Globals/ObjectMgr.h:877]: ???

**Comment by Warpten** [comment]
This is unnecessarily convoluted, why don't you just make that column an unique integer ?

**Comment by Seyden** [comment]
> This is unnecessarily convoluted, why don't you just make that column an unique integer ?

I have to update this PR but as we found out, StringID isn’t unique per creature.. multiple creature instances can have the same StringID.

The reason why it’s a string is basically to give the ID a certain key naming meaning

File: src/server/game/Globals/ObjectMgr.h (hunk 2 above)
Comment (Seyden) [reply src/server/game/Globals/ObjectMgr.h:877]: Any better idea how I can make it implicit castable to uint32?

Comment (Seyden) [reply src/server/game/Globals/ObjectMgr.h:877]: Naios proposed me to wrap it, I also don’t understand why this is good but well 🤷🏽‍♂️

**Comment by offl** [comment]
Could you please show how to use it in script?

**Comment by malcrom** [comment]
```
    void OnQuestStatusChange(Player* player, Quest const* /*quest*/, QuestStatus /*oldStatus*/, QuestStatus newStatus) override
    {
        uint32 index1 = sObjectMgr->GetStringIdIndex("kee_la_beach");
        uint32 index2 = sObjectMgr->GetStringIdIndex("bjorn_stouthands_beach");
        uint32 index3 = sObjectMgr->GetStringIdIndex("austin_huxworth_beach");

        switch (newStatus)
        {
            case QUEST_STATUS_INCOMPLETE:
                player->CastSpell(player, SPELL_SUMMON_ADMIRAL_GARRICK_GUARDIAN);
                if (Creature* keela = player->GetMap()->GetCreatureByStringId(index1))
                    keela->SummonPersonalClone(keela->GetPosition(), TEMPSUMMON_MANUAL_DESPAWN, 0s, 0, 0, player);
                if (Creature* bjorn = player->GetMap()->GetCreatureByStringId(index2))
                    bjorn->SummonPersonalClone(bjorn->GetPosition(), TEMPSUMMON_MANUAL_DESPAWN, 0s, 0, 0, player);
                if (Creature* austin = player->GetMap()->GetCreatureByStringId(index3))
                    austin->SummonPersonalClone(austin->GetPosition(), TEMPSUMMON_MANUAL_DESPAWN, 0s, 0, 0, player);
                break;
```

**Comment by Seyden** [comment]
> Could you please show how to use it in script?

Right now there isn’t a way, there is a different PR which I opened, I want to extend FindNearestCreatureWithOptions or create a new method for it.

In any case, the way to use it in the script is basically like this:

uint32 stringId = sObjectMgr->GetStringIdIndex("myStringId“);

and then you can pass this StringID in the certain find method or you can use it to write your own grid searcher and compare it with creature->GetStringIdIndex() == stringId

**Comment by offl** [comment]
Then this more looks like String, without ID
Some stuff about StringID is explained [here](https://docs.google.com/document/d/12H-IMOLv2Vp8yHjmNEH00Iy9WbK0LXswLErOYcj9Z7U/edit#heading=h.vooynpo17zek) and in comments on the right side

**Comment by Seyden** [comment]
> Then this more looks like String, without ID
> Some stuff about StringID is explained [here](https://docs.google.com/document/d/12H-IMOLv2Vp8yHjmNEH00Iy9WbK0LXswLErOYcj9Z7U/edit#heading=h.vooynpo17zek) and in comments on the right side

Well what you linked also proofed that the StringID can just be set to „onyxia“. This is exactly what you can do with this PR too. So I don’t understand your point :)

**Comment by Seyden** [comment]
We can now add a helper field for FindNearestCreatureWithOptions

**Comment by Seyden** [comment]
Updated description and some examples

**Pull request merged**